identity_stardust/src/client/client_ext.rs

```diff
@@ -0,0 +1,459 @@
+        QueryParameter::HasTimelockCondition(false),
+        QueryParameter::HasStorageReturnCondition(false),
+      ])
+      .await
+      .unwrap();
+
+    let outputs_responses = client.get_outputs(output_ids).await.unwrap();
+
+    let mut total_amount = 0;
+    for output_response in outputs_responses {
+      let output = Output::try_from(&output_response.output).unwrap();
+      total_amount += output.amount();
+    }
+
+    total_amount
+  }
+
+  async fn alias_output(client: &Client, address: Address, document: &StardustDocument) -> Output {
+    let rent_structure: RentStructure = client.get_rent_structure().await.map_err(Error::ClientError).unwrap();
+    AliasOutputBuilder::new_with_minimum_storage_deposit(rent_structure, AliasId::null())
+      .map_err(Error::AliasOutputBuildError)
+      .unwrap()
+      .with_state_index(0)
+      .with_foundry_counter(0)
+      .with_state_metadata(document.clone().pack().unwrap())
+      .add_feature(Feature::Sender(SenderFeature::new(address)))
+      .add_feature(Feature::Metadata(MetadataFeature::new(vec![1, 2, 3]).unwrap()))
+      .add_immutable_feature(Feature::Issuer(IssuerFeature::new(address)))
+      .add_unlock_condition(UnlockCondition::StateControllerAddress(
+        StateControllerAddressUnlockCondition::new(address),
+      ))
+      .add_unlock_condition(UnlockCondition::GovernorAddress(GovernorAddressUnlockCondition::new(
+        address,
+      )))
+      .finish_output()
+      .unwrap()
+  }
+
+  fn valid_did() -> StardustDID {
+    "did:stardust:0xAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+      .parse()
+      .unwrap()
+  }
+
+  async fn publish_document(
+    client: impl StardustClientExt,
+    document: StardustDocument,
+    address: Address,
+    secret_manager: &SecretManager,
+  ) -> crate::error::Result<(Block, StardustDocument)> {
+    let output = alias_output(client.client(), address, &document).await;
+
+    let block: Block = client.publish_outputs(secret_manager, vec![output]).await?;
+
+    let documents: Vec<StardustDocument> = client.documents_from_block(&block).await?;
+
+    Ok((
+      block,
+      documents
+        .into_iter()
+        .next()
+        .expect("there should be at least one document"),
+    ))
+  }
+
+  #[tokio::test]
+  async fn test_publish_resolve() {
+    let client: Client = client();
+    let (address, secret_manager) = get_address_with_funds(&client).await;
+    let document = generate_document(&valid_did());
+    let (_, document) = publish_document(&client, document, address, &secret_manager)
+      .await
+      .unwrap();
+    let resolved = client.resolve(document.id()).await.unwrap();
+
+    assert_eq!(document, resolved);
+  }
+
+  async fn publish_update(
+    client: &Client,
+    document: StardustDocument,
+    secret_manager: &SecretManager,
+  ) -> crate::error::Result<Block> {
+    let rent_structure: RentStructure = client.get_rent_structure().await.map_err(Error::ClientError)?;
```

Comment:
If we're managing updates at this level of abstraction, it might be convenient to allow specifying how to handle the `amount`?

E.g. an enum like

```rust
enum StorageAmount {
  Amount(u64), // try publish with the specified amount
  Minimum, // publish with the minimum deposit required
  #[default]
  Same/Previous, // do not alter the amount of tokens.
}
```

Reply:
I've added this method now:

```rust
/// Resolves the alias output associated to the DID in `document` and updates it with that document.
///
/// Returns the updated alias output for further customization and publication.
///
/// This method does not modify the on-ledger state.
async fn update_did(&self, document: StardustDocument, rent_structure: RentStructure)
-> Result<AliasOutput> { ... }
```

Do you think it's still necessary then? This method uses `with_minimum_storage_deposit` by default, but users can set the amount they want explicitly afterwards. This should overwrite the storage deposit I imagine:

```rust
let alias_output = client.update_did(...).await?;
AliasOutputBuilder::from(&alias_output).with_amount(1_000_000)?.finish()?;
```

This enables the functionality of your `Amount` variant, but not the `Same/Previous` variants, since the resolved output is hidden from the user so they cannot access the previous amount.

We could alternatively not use `with_minimum_storage_deposit` in `update_did`, so the deposit stays the same, and a user can use that method or `with_amount` to configure the functionalities you suggested. However, I don't think that's what the default should be, since that would oftentimes not cover the required deposit.

Reply:
Getting the `RentStructure` is annoying and should be handled by the method, or be optional to re-use a previous one.

I concur that `with_minimum_storage_deposit` should not be the default because it affects tokens. The `Same/Previous` behaviour seems the most sensible default, but updating to the minimum required storage deposit is very common and should be supported too. Again, it's just a case of how convenient we want the `Client` to be compared to the `Account`.

Reply:
Sorry for being unclear. I actually meant `with_minimum_storage_deposit` should be the default, because that's usually what's needed for the state update to succeed. Whether it affects tokens or not doesn't change how the API is called, so not sure where you see the issue.

Reply:
> Whether it affects tokens or not doesn't change how the API is called, so not sure where you see the issue.

The default should be to NOT affect people's tokens, anything potentially "dangerous" should be explicitly opt-in. The function signature of "update_did" shows no indication that it might take more funds or remove funds from the output (and documentation is insufficient per previous design decisions), hence an explicit parameter is preferred.